File: 62-unique-paths/unique-paths.ts
/*
 * @lc app=leetcode id=62 lang=typescript
 *
 * [62] Unique Paths
 */

// @lc code=start
function uniquePaths(m: number, n: number): number {
    
    const memo = Array.from({length: m}, () => Array(n).fill(-1));
    function getUniquePath(i: number, j: number): number{
        // base cases
        if (i === 0 && j === 0) return memo[i][j]=1;
        if(i < 0 || j < 0) return 0;
        if(memo[i][j]!==-1) return memo[i][j];
        // move up or left
        return memo [i][j]=getUniquePath(i-1, j) + getUniquePath(i, j-1);
    }
    // start from the goal 
    getUniquePath(m - 1, n - 1);
    return memo[m - 1][n - 1];
};
// @lc code=end

File: 62-unique-paths/unique-paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        memo = {}
        def f(r,c):
            if r==m-1 and c==n-1:
                return 1
            if r>=m or c>=n:
                return 0
            if (r,c) in memo:
                return memo[r,c]
            
            down = f(r+1,c)
            right = f(r,c+1)
            memo[r,c] = down+right
            return memo[r,c]
        return f(0,0)